AK/TaggedPtr.h

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2022, Jun Zhang <jun@junz.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Platform.h>
+#include <AK/Types.h>
+
+namespace AK {
+
+namespace detail {
+constexpr unsigned log2(unsigned bits)
+{
+    if (bits == 1)
+        return 0;
+
+    return log2(bits / 2) + 1;
+}
+}
+
+template<typename PtrType, unsigned Bits, typename TagType = unsigned>
+class TaggedPtr {
+    intptr_t value = 0;
```

Comment:
Also prefer `FlatPtr`

Reply:
Not sure about this, but open to discussion :)